tools/biosnoop.bt

```diff
@@ -1,43 +1,64 @@
 #!/usr/bin/env bpftrace
 /*
  * biosnoop.bt   Block I/O tracing tool, showing per I/O latency.
- *               For Linux, uses bpftrace, eBPF.
+ *	       For Linux, uses bpftrace, eBPF.
  *
- * TODO: switch to block tracepoints. Add device, offset, and size columns.
+ * TODO: switch to block tracepoints.
  *
  * This is a bpftrace version of the bcc tool of the same name.
  *
  * 15-Nov-2017	Brendan Gregg	Created this.
+ * 31-Jul-2020  Trent Lloyd     Added support for disk name, offset and length.
+ *       			Output systm time at startup. Requires bpftrace
+ *      			> 0.11.0 for more than 6 printf arguments.
  */
 
+#include <linux/blkdev.h>
+
 BEGIN
 {
-	printf("%-12s %-16s %-6s %7s\n", "TIME(ms)", "COMM", "PID", "LAT(ms)");
+	time("Start Time: %F %T\n");
+	printf("%-12s %-16s %-7s %-8s %-13s %-6s %7s\n", "TIME(ms)", "COMM", "PID", "DISK", "OFFSET", "LEN", "LAT(ms)");
 }
 
 kprobe:blk_account_io_start
 {
+	$req = ((struct request *)arg0);
+
 	@start[arg0] = nsecs;
 	@iopid[arg0] = pid;
 	@iocomm[arg0] = comm;
+	@iolen[arg0] = ((struct request *)arg0)->__data_len;
+	@iostart[arg0] = $req->__sector * $req->q->limits.logical_block_size
```

Comment:
Why stash these on start? Can't these fields be read on done, which would avoid needing to allocate the BPF maps? (The other fields were stashed on start since they are process context and time, and aren't available in the done arguments.)

Reply:
I originally tried that but the __data_len specifically gets zerod out during the request processing and reads 0 during blk_account_io_done. Looking at some kernel commits they renamed them with the private __ prefix in part because of some of that behaviour breaking external code relying on it but that's as far as I looked into that.

While the offset doesn't seem to do that at first glance I was concerned since it was the related private field it may also sometimes change and decided to stash that also -- but I didn't comprehensively look at the kernel paths to figure out if that was likely or not.